src/utils/Cache.ts

```diff
@@ -0,0 +1,246 @@
+import { AxiosRequestConfig, AxiosResponse } from 'axios';
+
+export type CacheTargets = CacheTarget[];
+export enum CacheTarget {
+  CACHE_API = 'CACHE_API',
+  MEMORY = 'MEMORY',
+}
+export type CacheResponse = {
+  data: any;
+  status: number;
+  statusText: string;
+  headers: any;
+};
+
+export const CACHE_API_KEY = 'sentinelhub-v1';
+export const SUPPORTED_TARGETS = [CacheTarget.CACHE_API, CacheTarget.MEMORY];
+const FALLBACK_TARGET = CacheTarget.MEMORY;
+
+export const memoryCache = new Map();
+
+// Factory will return an instance of a Cache interface.
+// The first availble target will be used
+// By default we will use cache_api if availble, if not we will cache in memory
+// user can also specify to just use in memory caching
+// If user provides a CacheTarget.CACHE_API and cache_api is not availble we will fallback to memory
+export function cacheFactory(optionalTargets: CacheTargets): ShCache {
+  const targets = optionalTargets || SUPPORTED_TARGETS;
+  const target = getFirstUsuableTarget(targets);
+  return constructCache(target);
+}
+
+function getFirstUsuableTarget(targets: CacheTargets): CacheTarget {
+  let firstTargetToUse = FALLBACK_TARGET; // default to memory if target is not supported
+  for (const key of targets) {
+    if (doesTargetExist(key)) {
+      firstTargetToUse = key;
+      break;
+    }
+  }
+  return firstTargetToUse;
+}
+
+export function doesTargetExist(target: CacheTarget): boolean {
+  switch (target) {
+    case CacheTarget.CACHE_API:
+      return Boolean(window.caches);
+    case CacheTarget.MEMORY:
+      return true;
+    default:
+      return false;
+  }
+}
+
+function constructCache(target: CacheTarget): ShCache {
+  switch (target) {
+    case CacheTarget.CACHE_API:
+      return new CacheApi();
+    case CacheTarget.MEMORY:
+      return new MemoryCache();
+    default:
+      throw new Error('Target not supported');
+  }
+}
+
+interface ShCache {
+  set(key: string, response: AxiosResponse): Promise<void>;
+  get(key: string, responseType: AxiosRequestConfig['responseType']): Promise<CacheResponse>;
+  has(key: string): Promise<boolean>;
+  keys(): Promise<string[]>;
+  delete(key: string): Promise<void>;
+  getHeaders(key: string): Promise<Record<string, any>>;
+  invalidate(): void;
+}
+
+class MemoryCache implements ShCache {
+  private cache: Record<string, any>;
+  public constructor() {
+    this.cache = memoryCache;
```

Comment:
Since we are always using an "external" (to class) resource to save cache data (`memoryCache`), and since we can't change `this.cache` anyway, I would suggest removing `this.cache` var and simply using `memoryCache` directly:
```
  public async set(key: string, response: AxiosResponse): Promise<void> {
    memoryCache.set(key, response);
  }
```

Reply:
Removed `this.cache` from `memoryCache` and `CacheApi`. Once it was removed from `CacheApi` another issue arose and that was that the return from `keys()` was incorrect, so this has also been fixed